packages/embark-compiler/src/index.ts

```diff
@@ -14,75 +16,166 @@ class Compiler {
     this.isCoverage = options.isCoverage;
 
     // embark.events.setCommandHandler("compiler:contracts", this.compile_contracts.bind(this));
-    embark.events.setCommandHandler("compiler:contracts:compile", this.compile_contracts.bind(this));
+    embark.events.setCommandHandler(
+      "compiler:contracts:compile",
+      this.compile_contracts.bind(this),
+    );
   }
 
-  private compile_contracts(contractFiles: any[], cb: any) {
+  private compile_contracts(contractFiles: any[], cb: Callback<any>) {
     if (contractFiles.length === 0) {
       return cb(null, {});
     }
 
-    const compiledObject: {[index: string]: any} = {};
+    async.waterfall(
+      [
+        (next: Callback<any[]>) => {
+          this.plugins.runActionsForEvent(
+            "compiler:contracts:compile:before",
+            contractFiles,
+            (err?: Error | null, files: any[] = []) => {
+              if (err) {
+                return next(err);
+              }
+              next(
+                null,
+                files.map((file: any) =>
+                  file instanceof File
+                    ? file
+                    : new File({
+                        originalPath: file.path,
+                        path: file.path,
+                        resolver: (callback: Callback<any>) => {
+                          fs.readFile(
+                            file.path,
+                            { encoding: "utf-8" },
+                            callback,
+                          );
+                        },
+                        type: Types.dappFile,
```

Comment:
I had to use the `dappFile` type even though that's not what's used here:

https://github.com/embark-framework/embark/blob/master/packages/embark/src/lib/core/config.js#L90-L95

Reply:
The reason for this setup is that otherwise a plugin needs to have `embark-utils` as a dependency, import/require File and Types, and build the `File` instance correctly. That seems like a lot for the plugin-author to have to manage; also, it makes it harder to have plugins that work with multiple major versions of embark because it could potentially involve the need to depend on multiple incompatible versions of `embark-utils`, which can't work. So I'm not sure what the best solution is.